### PR TITLE
fix: upload button position flashes when it changes from hidden to displayed

### PR DESCRIPTION
**Pull request by credred**
…o displayed (#33819)

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [x] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
https://github.com/ant-design/ant-design/issues/33819

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | fix the upload button sometimes flashes after the upload button is changed from hidden to display when the upload component is a picture-card type.         |
| 🇨🇳 Chinese | 修复Upload组件为picture-card类型时，由隐藏上传按钮改为显示后，上传按钮有时会闪烁的问题           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by credred** [comment]
PR为显示/隐藏上传按钮添加了动画，但是隐藏上传按钮时按钮区域内容会立即消失，是否考虑为`Upload`组件引入例如 `childrenVisible` 的属性来隐藏？

**Comment by MadCcc** [comment]
Some lint errors

**Comment by credred** [comment]
> Some lint errors

The lint error reports by stylelint for the value-keyword-case rule. I don't think  my commit is related to this error. Can you give me a suggestion i should solve this problem.

File: components/upload/Upload.tsx [hunk 1]
```diff
@@ -400,23 +401,23 @@ const InternalUpload: React.ForwardRefRenderFunction<unknown, UploadProps> = (pr
     [`${prefixCls}-rtl`]: direction === 'rtl',
   });
 
-  const uploadButton = (
-    <div className={uploadButtonCls} style={children ? undefined : { display: 'none' }}>
+  const renderUploadButton = (uploadButtonStyle?: React.CSSProperties) => (
```
Comment (MadCcc) [inline components/upload/Upload.tsx:404]: 这个改动看上去没什么用？

File: components/upload/UploadList/index.tsx [hunk 2]
```diff
@@ -225,12 +226,14 @@ const InternalUploadList: React.ForwardRefRenderFunction<unknown, UploadListProp
...
       {/* Append action */}
       {appendAction && (
-        <CSSMotion {...motionConfig}>
+        <CSSMotion {...motionConfig} visible={appendActionVisible} forceRender>
           {({ className: motionClassName, style: motionStyle }) =>
             cloneElement(appendAction, oriProps => ({
               className: classNames(oriProps.className, motionClassName),
               style: {
                 ...motionStyle,
+                // prevent the element has hover css pseudo-class that way cause animation to end prematurely.
```
Comment (MadCcc) [inline components/upload/UploadList/index.tsx:235]:
```suggestion
                // prevent the element has hover css pseudo-class that may cause animation to end prematurely.
```

Comment (MadCcc) [inline components/upload/UploadList/index.tsx:235]:
```suggestion
                // prevent the element has hover css pseudo-class that may cause animation to end prematurely.
```

File: components/upload/UploadList/index.tsx [hunk 3]
```diff
@@ -225,12 +226,14 @@ const InternalUploadList: React.ForwardRefRenderFunction<unknown, UploadListProp
 
       {/* Append action */}
       {appendAction && (
-        <CSSMotion {...motionConfig}>
+        <CSSMotion {...motionConfig} visible={appendActionVisible} forceRender>
```
Comment (MadCcc) [inline components/upload/UploadList/index.tsx:229]: forceRender 是必要的吗

File: components/upload/Upload.tsx (hunk 1 above)
Comment (credred) [reply components/upload/Upload.tsx:404]: 普通的UploadButton仍然保留直接隐藏，listType为picture-card的UploadButton将隐藏功能交给UploadList，在动画结束后才隐藏

**Comment by MadCcc** [comment]
> PR为显示/隐藏上传按钮添加了动画，但是隐藏上传按钮时按钮区域的内容会立即消失，是否考虑为`Upload`组件引入例如 `childrenVisible` 的属性来隐藏或者缓存最后一次非null的children？

issue 里的显示问题是用什么方式解决的呢

File: components/upload/UploadList/index.tsx (hunk 3 above)
Comment (credred) [reply components/upload/UploadList/index.tsx:229]: 不加forceRender在结束后会销毁AppendAction(也就是UploadButton)。可能会导致如[16478](https://github.com/ant-design/ant-design/issues/16478)的问题：上传过程中销毁了Upload组件。

**Comment by credred** [comment]
> > PR为显示/隐藏上传按钮添加了动画，但是隐藏上传按钮时按钮区域的内容会立即消失，是否考虑为`Upload`组件引入例如 `childrenVisible` 的属性来隐藏或者缓存最后一次非null的children？
> 
> issue 里的显示问题是用什么方式解决的呢

在隐藏上传按钮时为其添加移出动画，也就是`uploadAnimateInlineOut`。我来尝试提供下解决后的重现

**Comment by MadCcc** [comment]
> The lint error reports by stylelint for the value-keyword-case rule. I don't think my commit is related to this error. Can you give me a suggestion i should solve this problem.

合一下 master 最新的代码

**Comment by credred** [comment]
> > > PR为显示/隐藏上传按钮添加了动画，但是隐藏上传按钮时按钮区域的内容会立即消失，是否考虑为`Upload`组件引入例如 `childrenVisible` 的属性来隐藏或者缓存最后一次非null的children？
> > 
> > 
> > issue 里的显示问题是用什么方式解决的呢
> 
> 在隐藏上传按钮时为其添加移出动画，也就是`uploadAnimateInlineOut`。我来尝试提供下解决后的重现

https://codesandbox.io/s/pr-ant-design-upload-animation-0bf2k

**Pull request merged**